### ISSUE TITLE
Modal - allow aria-labeldBy prop to be passed

### ISSUE THREAD

**Issue by alexandrzavalii** [Workday/canvas-kit#756]
## 🚀 Feature Proposal

We are using Modal with a custom title, and we need to pass down aria-labeldBy to the dialog.
```
<Modal aria-labeldby="title">
  <h1 id="title">Title</h1>
</Modal>
```

Currently the modal accepts only `heading` prop which will create a `h3` title, and `aria-label`, but we are missing aria-labeldby to target our custom title.

**Comment by mathalete** [comment]
This is to provide 2 means to add an accessible name to a modal.

What we need here is to account for modals that have a visible heading, and those that don't.

### When there is a visible heading

The component needs to allow adding the aria-labelledby attribute to the modal container, and pointing the value of this to the ID of the heading of the modal.

Code sample

```
<div aria-modal="true" aria-labelledby="h1">
 <h1 id="h1>Title</h1>
...
</div>
```

### When there is no visual heading

When there is no visible heading in the modal, an aria-label is used to provide the accessible name.

Code sample

`<div aria-modal="true" aria-label"My big fat hairy modal">...</div>`

**Comment by NicholasBoll** [comment]
Related to #493

**Comment by NicholasBoll** [comment]
This is on our radar to align with Modal specs: https://www.figma.com/file/IjDja6ooe8B4JTiMz8VJU4/%F0%9F%9B%A0-Popups---Modal?node-id=76%3A0&viewport=66%2C1459%2C0.6816627383232117

We'll need a way to support custom headers. `h3` was chosen arbitrarily and I don't think this makes sense. Since everything else is hidden on the page, an `h1` seems more appropriate (the entire document outline is already changed when the modal is active).

For accessibility, we still need some way of enforcing Modals to be labelled. Even if perhaps only by a runtime warning (Typescript doesn't handle "either this prop or this prop needs to be included, but not both" very well).

The proposal as-is will not work due to strict requirements of focus management. We need to know what the intended header should be. In the case of no close icon, the focus is shifted to the heading for screen readers to start reading content from the start of the modal. Without tight control over the header, this functionality becomes fragile. Right now an active error will be thrown if there is nothing focusable in the modal.

I do agree that we need some way to customize header content and handle a non-visible header.

**Comment by NicholasBoll** [comment]
@mathalete would you mind looking at #493 and commenting? Is it very possible we close that issue in favor of this one.

**Comment by mathalete** [comment]
@NicholasBoll As it's a change of context, a <h1> is recommended if there is a visible heading.

I can't at the moment think of a use case where there'll be a modal with nothing focusable inside of it - there should always be a cancel or close button - otherwise a keyboard user would need to guess pressing the escape key will dismiss it.

There should be flexibility for modals - or just a default rule that a modal must have visible heading as a design system pattern - this is the best approach.

**Comment by mathalete** [comment]
@NicholasBoll As it's a change of context, a <h1> is recommended if there is a visible heading. 

I can't at the moment think of a use case where there'll be a modal with nothing focusable inside of it - there should always be a cancel or close button - otherwise a keyboard user would need to guess pressing the escape key will dismiss it. 

There should be flexibility for modals - or just a default rule that a modal must have visible heading as a design system pattern - this is the best approach.

Will look at #493

**Comment by NicholasBoll** [comment]
> I can't at the moment think of a use case where there'll be a modal with nothing focusable inside of it - there should always be a cancel or close button - otherwise a keyboard user would need to guess pressing the escape key will dismiss it.

That statement will be a debate with design. Most modal dialogs will be easily dismissible via a close icon. But, in the past I've come across cases where modals contain forms with work a user doesn't want to lose unless they are explicitly deciding to. An `escape` key can be too easy to hit accidentally if work will be lost (i.e. escape key dismisses an open select menu or tooltip, but one extra escape closes the modal and possibly loses work). UX may request a modal to be explicitly closable via a "Cancel" or "I don't agree". The modal currently supports those cases by making the header be focused at the request of the accessibility team (the idea being that the modal's contents will be readable from the top). If the "Cancel" button is required instead, that will also be a debate. We also have to consider that the Modal doesn't own any internal buttons - only the optional close icon and header. Any changes to those requirements means a much looser accessibility contract that the developers now have the burden to fulfill.

I think a requirement like "A Modal requires _either_ an `aria-label` _or_ a `header` (which contents will be put in an `h1`)".  Right now we use the `header` to seamlessly support accessibility concerns. To much flexibility means we cannot make that seamless. Perhaps we really want that flexibility, but then we'll have to document "and these are the extra things you have to think about".

**Comment by NicholasBoll** [comment]
In v5, this is now possible:

```tsx
<Modal>
  <Modal.Target as={DeleteButton}>Delete Item</Modal.Target>
  <Modal.Overlay>
    <Modal.Card aria-labelledby="my-heading">
      <Modal.CloseIcon aria-label="Close" />
      <h1 id="my-heading">Modal Title</h1>
      <Modal.Body>
        ...
      </Modal.Body>
    </Modal.Card>
  </Modal.Overlay>
</Modal>
```

Although this is not recommended. The `Modal.Heading` does additional accessibility work, but can appear anywhere in the Modal. The `aria-labelledby` and `id` are already handled by the `Modal` and subcomponents. Use `<Modal.Heading as="h1">` instead of `<h1 id="...">`.
